### PR TITLE
Add new factories 

### PR DESCRIPTION
**Pull request by harikt**

### CODE REVIEW

File: src/HelperLocator.php [hunk 1]
```diff
@@ -3,11 +3,11 @@
 
 class HelperLocator
 {
-    static public function new(Escape $escape = null) : self
+    static public function new(Escape $escape = null, $factories = []) : self
```
Comment (koriym) [inline src/HelperLocator.php]:
```suggestion
    static public function new(Escape $escape = null, array $factories = []) : static
```

Comment (koriym) [reply src/HelperLocator.php]: for the inherited class

**Review by koriym — CHANGES_REQUESTED** [review]

**Comment by koriym** [comment]
I think this implementation is preferable to the current one where the state of the HelperLocator object is changed later by a "setter".

File: src/HelperLocator.php (hunk 1 above)
Comment (harikt) [reply src/HelperLocator.php]: I think your suggestion has a typo. `staic` vs `static`

Comment (harikt) [reply src/HelperLocator.php]: Another note probably is type hint to array. So I can have both in single commit. Thougths ?

Comment (koriym) [reply src/HelperLocator.php]: Sure. Updated!

File: src/HelperLocator.php [hunk 2]
```diff
@@ -3,11 +3,11 @@
 
 class HelperLocator
 {
-    static public function new(Escape $escape = null) : self
+    static public function new(Escape $escape = null, array $factories = []) : static
```
Comment (harikt) [inline src/HelperLocator.php:6]: Hey @koriym ,
I am wondering what is the change from self to static to do. Also this one is returning self. So the return type should also be self right ?

File: src/HelperLocator.php [hunk 3]
```diff
@@ -55,7 +55,11 @@ static public function new(Escape $escape = null) : self
             'ul'                    => function () use ($escape) { return new Helper\Ul($escape); },
             'urlField'              => function () use ($escape) { return new Helper\UrlField($escape); },
             'weekField'             => function () use ($escape) { return new Helper\WeekField($escape); },
-        ]);
+        ];
+
+        $helper_factories = array_merge($default, $factories);
+
+        return new self($helper_factories);
```
Comment (koriym) [inline src/HelperLocator.php]:
```suggestion
        return new static($helper_factories);
```

Comment (koriym) [reply src/HelperLocator.php]: You need to make it static here as well.

**Review by koriym — CHANGES_REQUESTED** [review]
```php
class A
{
    static public function new(): self
    {
        return new self();
    }

    public function get()
    {
        return 'service locater';
    }
}

class B extends A
{
    public function get()
    {
        return 'lazy-load';
    }
}

var_dump(B::new()->get()); // service locater
```

If we want to override a method in an inherited class, we can't do it with "self".

**Comment by pmjones** [comment]
Nice! Thanks for your patience, gentlemen.

**Pull request merged**